File: mpp/Lab6/mpp-proiect-java/javaPb9/src/main/java/ro/mpp2024/repository/ParticipantRepository.java
package ro.mpp2024.repository;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import ro.mpp2024.model.Participant;
import ro.mpp2024.repository.interfaces.IParticipantRepository;
import ro.mpp2024.utils.JdbcUtils;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class ParticipantRepository implements IParticipantRepository {

    private JdbcUtils dbUtils;

    private static final Logger logger= LogManager.getLogger();


    public ParticipantRepository(JdbcUtils dbUtils) {
        logger.info("Initializing ParticipantRepository with properties: {} ");
        this.dbUtils = dbUtils;
    }

    @Override
    public Participant createEntityFromResultSet(ResultSet resultSet) throws SQLException {
        logger.traceEntry("createEntityFromResultSet");
        Long id = resultSet.getLong("id");
        String nume = resultSet.getString("nume");
        String prenume = resultSet.getString("prenume");
        int varsta = resultSet.getInt("varsta");
        Participant participant = new Participant(id,nume,prenume,varsta);
        logger.traceExit(participant);
        return participant;
    }

    @Override
    public Iterable<Participant> findAll() {
        logger.traceEntry();
        Connection con=dbUtils.getConnection();
        List<Participant> participants = new ArrayList<>();
        try(PreparedStatement preStmt=con.prepareStatement("select * from Participant")){
            try(ResultSet result=preStmt.executeQuery()){
                while(result.next()){
                    Participant participant = createEntityFromResultSet(result);
                    participants.add(participant);
                }
            }
        }catch (Exception ex){
            logger.error(ex);
            System.out.println("Error DB "+ex);
        }
        logger.traceExit(participants);
        return participants;
    }

    @Override
    public Participant findOne(Long id) {
        logger.traceEntry();
        Connection con=dbUtils.getConnection();
        Participant participant = null;
        try(PreparedStatement preStmt=con.prepareStatement("select * from Participant where id=?")) {
            preStmt.setLong(1, id);
            try (ResultSet result = preStmt.executeQuery()) {
                if (result.next()) {
                    participant = createEntityFromResultSet(result);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error DB " + e);
        }
        logger.traceExit(participant);
        return participant;
    }

    @Override
    public void save(Participant entity) {
        logger.traceEntry("save");
        Connection con=dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("insert into Participant(nume,prenume,varsta) values (?,?,?)")){
            preStmt.setString(1,entity.getNume());
            preStmt.setString(2,entity.getPrenume());
            preStmt.setInt(3,entity.getVarsta());
            int result=preStmt.executeUpdate();
        }catch (SQLException ex){
            logger.error(ex);
            System.out.println("Error DB "+ex);
        }
        logger.traceExit("S-a adaugat");
    }

    @Override
    public Participant delete(Long aLong) {
        //TODO: implement method when neeeded
        return null;
    }

    @Override
    public Participant update(Long aLong, Participant new_entity) {
        //TODO: implement method when neeeded
        return null;
    }
}

File: mpp/Lab10/csharpPb9/persistene/repository/ParticipantRepository.cs
using csharpPb9.utils;
using log4net;
using System.Data;

public class ParticipantRepository : IParticipantRepository
{
    private static readonly ILog log = LogManager.GetLogger("");

    IDictionary<string, string> props;

    public ParticipantRepository(IDictionary<string, string> props)
    {
        log.Info("Creating ArbitruRepository with properties: " + props);
        this.props = props;
    }

    public Participant fromSettoEntity(IDataReader read)
    {
        long id = read.GetInt64(0);
        string nume = read.GetString(1);
        string prenume = read.GetString(2);
        int varsta = read.GetInt32(3);

        return new Participant(id, nume, prenume, varsta);
    }

    public IEnumerable<Participant> FindAll()
    {
        log.Info("Finding all participanti");
        IDbConnection con = DBUtils.getConnection(props);
        IList<Participant> participanti = new List<Participant>();
        try
        {
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select * from Participant";
                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        Participant participant = fromSettoEntity(dataR);
                        participanti.Add(participant);
                    }
                }
            }
        }
        catch (Exception e)
        {
            log.Error(e);
            Console.WriteLine(e.Message);
        }
        finally
        {
            log.Info("Closing connection");
            con.Close();
        }

        log.InfoFormat("Exiting FindAll with value {0}", participanti);
        return participanti;
    }

    public Participant FindOne(long id)
    {
        log.Info("Finding One participant");
        IDbConnection con = DBUtils.getConnection(props);
        Participant participant = null;
        try
        {
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "Select * from Participant where id=@id";
                var param = comm.CreateParameter();
                param.ParameterName = "@id";
                param.Value = id;
                comm.Parameters.Add(param);
                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        participant = fromSettoEntity(dataR);
                    }
                }
            }
        }
        catch (Exception e)
        {
            log.Error(e);
            Console.WriteLine(e.Message);
        }
        finally
        {
            log.Info("Closing connection");
            con.Close();
        }

        log.InfoFormat("Exiting FindOne with value {0}", participant);
        return participant;
    }


    public void Save(Participant entity)
    {
        log.Info("Save Participant");
        IDbConnection con = DBUtils.getConnection(props);

        try
        {
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "insert into Participant(nume,prenume,varsta) values (@nume, @prenume, @varsta)";
                #region vars
                var numeParam = comm.CreateParameter();
                numeParam.ParameterName = "@nume";
                numeParam.Value = entity.Nume;
                comm.Parameters.Add(numeParam);

                var prenumeParam = comm.CreateParameter();
                prenumeParam.ParameterName = "@prenume";
                prenumeParam.Value = entity.Prenume;
                comm.Parameters.Add(prenumeParam);

                var varstaParam = comm.CreateParameter();
                varstaParam.ParameterName = "@varsta";
                varstaParam.Value = entity.Varsta;
                comm.Parameters.Add(varstaParam);
                #endregion
                var result = comm.ExecuteNonQuery();
                if (result == 0)
                {
                    log.Error("No arbitru saved!");
                }
                else
                {
                    log.Info("Arbitru saved!");
                }
            }
        }
        catch (Exception e)
        {
            log.Error(e);
            Console.WriteLine(e.Message);
        }
        finally
        {
            log.Info("Closing connection");
            con.Close();
        }

        log.Info("Exiting Save");
    } 
    #region TO implement
    public Participant Delete(long id)
    {
        throw new NotImplementedException();
    }

    public Participant Update(long id, Participant new_entity)
    {
        throw new NotImplementedException();
    }
    #endregion 
}
